### PR TITLE
Json authenticator support mTLS username/client certificate verification

### PR DESCRIPTION
**Pull request by robklg**
https://github.com/bolcom/libunftp/issues/369

### CODE REVIEW

**Comment by robklg** [comment]
This implementation does no longer allow for any client certificate to be used for any user account. I guess that makes sense right? Otherwise we'd need to provide a global feature to enable that possibility but it doesn't seem logical at all to me.

File: src/auth/authenticator.rs [hunk 1]
```diff
@@ -95,6 +99,21 @@ impl From<&str> for Credentials {
...
+impl ClientCert {
+    /// Returns true if the Common Name from the client certificate matches the allowed_cn
+    pub fn verify_cn(&self, allowed_cn: &str) -> Result<bool, std::io::Error> {
+        let client_cert = parse_x509_certificate(&self.0);
+        let subject = match client_cert {
+            Ok(c) => c.1.subject().to_string(),
+            Err(e) => return Err(std::io::Error::new(std::io::ErrorKind::Other, e.to_string())),
+        };
+
+        Ok(subject.contains(allowed_cn))
```
Comment (hannesdejager) [inline src/auth/authenticator.rs:113]: should it be contains or equals?

File: crates/unftp-auth-jsonfile/src/lib.rs [hunk 2]
```diff
@@ -329,5 +425,237 @@ mod test {
...
+  {
+    "username": "carol",
+    "client_cert": {
+      "allowed_cn": "unftp-other-client.mysite.com"
+    }
+  },
+  {
+    "username": "dean",
+    "client_cert": {}
+  }
```
Comment (hannesdejager) [inline crates/unftp-auth-jsonfile/src/lib.rs:576]: Perhaps good to also add the case where there is an empty client cert object and also a password i.e:

```json
  {
    "username": "santa",
    "password": "clara",
    "client_cert": {}
  }
```

**Comment by hannesdejager** [comment]
Please rebase.

File: src/auth/authenticator.rs (hunk 1 above)
Comment (robklg) [reply src/auth/authenticator.rs:113]: contains because the subject holds the DN

**Pull request merged**